### DISCUSSION TITLE
Dev

### DISCUSSION THREAD

**Discussion by cdnninja** [cdnninja/yoto_api#5] [General]
@buzzeddesign  starting a discussion to share progress

**Comment by cdnninja** [comment]
Do you have auth working locally in post man? I saw you had a few auth related URLs in the details.   If so we could skip using the 0auth library and just mimic what you have.  If not I can keep playing with auth0. I don't have much to go on with that though.

**Reply by buzzeddesign** [reply]
Haven't tried locally yet (will try now), but suspect it'll work with the bearer token!  The bearer token needs to be periodically refreshed and not sure how that is typically managed in Home Assistant (glad you know that part ;-)

Just added PR with what I think might work in the `requests.get` command

**Reply by buzzeddesign** [reply]
Confirmed this works locally if you have the bearer token:

```curl -H $auth https://api.yotoplay.com/card/family/library | jq```

**Reply by cdnninja** [reply]
Perfect yes can get the HA integration going.    

I am thinking once basic data is working in this API package I will publish it and frame up the HA plugin.  Waiting on that to first confirm we can get data.

**Reply by buzzeddesign** [reply]
Getting the bearer token is straightforward - 

```
$ pass='audience=https%3A//api.yotoplay.com&client_id=FILL_THIS_IN&grant_type=password&password=FILL_THIS_IN&scope=openid%20email%20profile%20offline_access&username=FILL_THIS_IN%40gmail.com'
$ curl -d "$pass" https://api.yotoplay.com/auth/token | jq '.access_token'
```

The auth/token URL also returns a refresh token - that refresh token can be used to get a new bearer token (not exactly sure when this needs to be done):

```
audience=https%3A//api.yotoplay.com&client_id=FILL_THIS_IN&grant_type=refresh_token&refresh_token=FILL_THIS_IN&scope=openid%20email%20profile%20offline_access
```

**Reply by cdnninja** [reply]
Perfect! I will make a few tweaks right now in my PR to align to this.

**Reply by cdnninja** [reply]
Merged #10  Take a look and let me know if it makes sense.   I think headers for token will need changes but will need your input on that.   I haven't run that code yet.

**Reply by cdnninja** [reply]
Alright.  I have done a test run of the code that is currently checked in.  

Right now it fails on login with the response: 

2024-04-13 20:52:49,233 yoto_api.YotoAPI DEBUG:yoto_api - Sign In Response {'error': 'server_error', 'error_description': 'Something went wrong on our side :('}

So not sure why it doesn't like it but something is missing.  Maybe headers on the token URL?

**Reply by buzzeddesign** [reply]
I was getting that as well - think I've resolved in the latest commit! See if this is working for you as well

**Reply by cdnninja** [reply]
Nice work!  Works for me here too.  Now have a populated token.    Next up will be flushing out some of this data and connecting it up with yoto manager.

**Reply by cdnninja** [reply]
Looks like authentication is working.  I have the yotoplayer objects populating.    If you have cycles and want to help out feel free to add other methods and data mapping following the approach of device.   

I am also missing other endpoints. I have devices for example but not the status of the device and settings etc.

**Reply by cdnninja** [reply]
Thats probably all I will get to today. https://github.com/cdnninja/yoto_ha has been created and I did the first bit to publish this api to pypi. 

I haven't run the HA code yet to see if it runs.    I suspect it will error.   Right now it is written to create a device and have a single sensor.  That sensor data needs to be added to the mapping still.   Once that is all working we essentially just need to add features.

**Reply by cdnninja** [reply]
Home assistant integration now starts, can be added with no errors.  Only sensor is the "online" data point.  However that is a good start.

**Reply by bushbrother** [reply]
I gave it a whirl (have been following this dev work - you are much faster than me!). I get login errors in HA, I double checked and I am using the right values for user/pass.

**Reply by cdnninja** [reply]
You may have caught between my broken commits.    I added another sensor type in HA so that is ready to go too.   Currently shows just two data points.  Online status and last updated.   Enough to prove out the structure is loading. 

I could really use a hand sniffing to see what other endpoints are used. In particular the one that shows what is playing, light color etc. I assume a device details endpoint exist that I can parse a ton of data from.

After that the "play" endpoint is of interest.  

Not sure if any of these are mqtt or not.

**Reply by bushbrother** [reply]
I removed it and re-downloaded, still not working ... I get "Login failed into Yoto Servers. Please use official app to logout and log back in and try again!"

**Reply by cdnninja** [reply]
Could you enable debug logs and provide those?

To confirm you are using the login to the mobile app not the community website?

**Reply by bushbrother** [reply]
I confirm I am using the credentials from the app. Not sure how to enable debug logs as the service is not created as it cannot login, so nowhere to enable :(

**Reply by cdnninja** [reply]
Maybe check if any logs in system logs to see if anything helpful on that front? 

If that doesn't work we can try run the python directly to see what is happening from the api side.

**Reply by bushbrother** [reply]
Nothing in the logs sadly, I had checked before. How can I run the python from within HA directly?

**Reply by cdnninja** [reply]
It won't be within HA.  

The readme on this repo has the code of how I test.  Download the repo, ideally with GitHub desktop.   Copy paste that contents into another file somewhere else on the pc.   

Correct the path to the api and enter creds and run it.

**Reply by bushbrother** [reply]
will try this evening if i get a chance. Thanks.

**Reply by bushbrother** [reply]
OK got round to this. Firstly your code had issues for me. There is a \ before the * on line 8 that is not required and for me logger needed to be as __ name __ not with *.

Anyway, long story short I used the same credentials and it logs in fine and i see all the card data etc.

**Reply by cdnninja** [reply]
I have made corrections to the docs.  Not sure how that happened.   

I also made some updates today.  Possible fixing what was happening for you but doubtful.   I think we move this to an issue in the _HA repo if still an issue. Could you retry?

**Comment by cjb0001** [comment]
Here are the endpoints I'm aware of:

GET APIs
https://api.yotoplay.com/device-v2/devices/mine
https://api.yotoplay.com/device-v2/$deviceid/status
https://api.yotoplay.com/media/displayIcons/user/me
https://api.yotoplay.com/user/details
https://api.yotoplay.com/user/family/mine?allowStub=true
https://api.yotoplay.com/card/mine
https://api.yotoplay.com/card/mine/user/family/mine?allowStub=true
https://api.yotoplay.com/card/family/library
https://api.yotoplay.com/card/library/free
https://api.yotoplay.com/card/library/club
https://api.yotoplay.com/card/family/library

Please note, as I stated in the HA Integration repo, I feel using MQTT is the MUCH BETTER way to get status instead of hammering the status endpoint.

**Reply by cdnninja** [reply]
Thanks for providing this.  I almost get the impression you work for yoto with the knowledge you have! Greatly appreciated.

In concept agreed on MQTT.  However I don't see certain data coming back around the player when I watch MQTT subscribed to the 3 topics.  An example is device status data.  It appears this only comes from the API above rather than MQTT.  MQTT seems to offer just playback related items.   Does this align to your understanding?

**Reply by cdnninja** [reply]
I suspect some post APIs exist for certain config items.

**Reply by cdnninja** [reply]
Also would love support. Feel free to create PRs on this if you are interested and have time. I am thinking each of the above APIs need to be added to the API file and mapped to any relevant data.

**Comment by cjb0001** [comment]
> I don't see certain data coming back around the player when I watch MQTT subscribed to the 3 topics. An example is device status data.

From the mqtt /status topic I see quite a bit of information.  I haven't done a comprehensive cross-reference vs the $deviceid/status API but it looks like all data is present in mqtt. 

{
    "status": {
        "activeCard": "XXXXX",
        "aliveTime": 2527630,
        "als": 0,
        "battery": 4088,
        "batteryData": "0:0:0",
        "batteryLevel": 100,
        "batteryLevelRaw": 99,
        "batteryProfile": "JY734352",
        "batteryTemp": 0,
        "bgDownload": 0,
        "bluetoothHp": 0,
        "buzzErrors": 0,
        "bytesPS": 0,
        "cardInserted": 1,
        "charging": 0,
        "chgStatLevel": 0,
        "day": 1,
        "dayBright": 100,
        "dbatTimeout": 60,
        "dnowBrightness": 100,
        "errorsLogged": 6,
        "failData": 0,
        "failReason": 0,
        "free": 3831100,
        "free32": 66190,
        "freeDisk": 6406112,
        "freeDMA": 56202,
        "fwVersion": "v2.17.6",
        "headphones": 0,
        "missedLogs": 0,
        "nfcErrs": "0.00%-n/a",
        "nightBright": 100,
        "nightlightMode": "off",
        "playingStatus": 0,
        "powerCaps": "",
        "powerSrc": 0,
        "productType": "mini",
        "qiOtp": 0,
        "sd_info": "0xad, 0x4c53(LS), USD00, 16, 896843272, 11-2022, 1, 0, 1461, 50000000",
        "shutDown": "nA",
        "shutdownTimeout": 1800,
        "ssid": "XXXXXXXX",
        "statusVersion": 3,
        "temp": "0:0:0",
        "timeFormat": "24",
        "totalDisk": 15441920,
        "twdt": 0,
        "upTime": 10,
        "userVolume": 50,
        "utcOffset": 7200,
        "utcTime": 1713872497,
        "volume": 55,
        "wifiRestarts": 1,
        "wifiStrength": -34
    }
}

**Reply by cdnninja** [reply]
Interesting.  I had put my status MQTT response here: https://github.com/cdnninja/yoto_api/blob/41312ace71a7d373fbfb13ff838d5f830bc15c13/yoto_api/YotoPlayer.py#L39-L57

Maybe it changes based on what is happening.

**Reply by cjb0001** [reply]
Yes, it seems to update when things are happening. I turned it on and off, played, paused, took cards in and out, and the topic fills up nicely with information.

**Comment by cdnninja** [comment]
@cjb0001 I have mqtt working with HA now.     Need to dial it in for update speed but data is flowing.  

I don't have the endpoint though for setting light status on the yoto.  Any chance you could sniff the control and status API endpoint? it doesn't seem to go across MQTT.

**Reply by cdnninja** [reply]
I also would love to know how you are collecting these endpoints.

**Reply by cjb0001** [reply]
Nice work getting MQTT going!! I'll take a look at it soon. 

I only have a mini so can't confirm any of this color stuff, but here's some info I was able to glean for you:

hex | color
-- | --
#40bfd9 | sky_blue
#41c0f0 | sky_blue
#9eff00 | apple_green
#e6ff00 | apple_green
#f57399 | lilac
#f72a69 | lilac
#ff0000 | tambourine_red
#ff3900 | orange_peel
#ff8500 | bumblebee_yellow
#ff8c00 | orange_peel
#ffb800 | bumblebee_yellow
#ffffff" | white
#0 | black ambient_light_off


For setting the light, it looks to be:
METHOD: GET
URL:    https://api.yotoplay.com/device-v2/xxDEVICEIDxx/config

METHOD: PUT
URL:    https://api.yotoplay.com/device-v2/xxDEVICEIDxx/config
REQUEST BODY:
{
  "deviceId": "xxDEVICEIDxx",
  "name": "WXYZ’s Yoto Player",
  "fwVersion": "v2.17.7",
  "popCode": "XXXXXX",
  "releaseChannelId": "beta",
  "activationPopCode": "XXXXXXXX",
  "deviceType": "mini",
  "deviceFamily": "mini",
  "deviceGroup": "",
  "mac": "xx:xx:xx:xx:xx:xx",
  "online": false,
  "geoTimezone": "My/Loc",
  "config": {
    "bluetoothEnabled": "1",
    "showDiagnostics": true,
    "displayDimTimeout": "60",
    "shutdownTimeout": "1800",
    "headphonesVolumeLimited": true,
    "repeatAll": false,
    "dayTime": "07:15",
    "maxVolumeLimit": "13",
    "ambientColour": "#ff0000",
    "dayDisplayBrightness": "99",
    "dayYotoDaily": "xxxxx/daily/<yyyymmdd>",
    "dayYotoRadio": "xxxxx/radio-day/01",
    "daySoundsOff": "0",
    "nightTime": "20:00",
    "nightMaxVolumeLimit": "12",
    "nightAmbientColour": "#ff0000",
    "nightDisplayBrightness": "100",
    "nightYotoDaily": "xxxxx/daily/<yyyymmdd>",
    "nightYotoRadio": "xxxxx/radio-night/01",
    "nightSoundsOff": "1",
    "hourFormat": "24",
    "timezone": "",
    "displayDimBrightness": "0",
    "systemVolume": "100",
    "volumeLevel": "safe",
    "clockFace": "digital-sun",
    "logLevel": "none",
    "alarms": []
  }
}

**Reply by cdnninja** [reply]
Those colors looks to be a different format than I get back.   I have the larger v3 player though.

**Reply by cjb0001** [reply]
> I also would love to know how you are collecting these endpoints.

On the API side, Android Studio, the Yoto APK, and HTTP Toolkit.  Bit of Node-Red for various digging too. On the mqtt side, MQTT Explorer.
